_posts/wasmer-go-embedding-1.0.md

```diff
@@ -0,0 +1,534 @@
+// (for the sake of simplicity of the example).
+module, _ := wasmer.NewModule(
+	store,
+	[]byte(`
+		(module
+		  ;; We import a math.sum function.
+		  (import "math" "sum" (func $sum (param i32 i32) (result i32)))
+
+		  ;; We export an add_one function.
+		  (func (export "add_one") (param $x i32) (result i32)
+		    local.get $x
+		    i32.const 1
+		    call $sum))
+	`),
+)
+
+// Let's create a new host function for `math.sum`.
+function := wasmer.NewFunction(
+	store,
+
+	// The function signature.
+	wasmer.NewFunctionType(
+		// Parameters.
+		wasmer.NewValueTypes(wasmer.I32, wasmer.I32),
+		// Results.
+		wasmer.NewValueTypes(wasmer.I32)
+	),
+
+	// The function implementation.
+	func(args []wasmer.Value) ([]wasmer.Value, error) {
+		x := args[0].I32()
+		y := args[1].I32()
+
+		return []wasmer.Value{wasmer.NewI32(x + y)}, nil
+	},
+)
+
+// Let's use the new `ImportObject` API…
+importObject := wasmer.NewImportObject()
+
+// … to register the `math.sum` function.
+importObject.Register(
+	"math",
+	map[string]wasmer.IntoExtern{
+		"sum": function,
+	},
+)
+
+// Finally, let's instantiate the module, with all the imports.
+instance, _ := wasmer.NewInstance(module, importObject)
+
+// And let's call the `add_one` function!
+addOne, _ := instance.Exports.GetFunction("add_one")
+
+result, _ := addOne(41)
+
+assert(result, int32(42))
+```
+
+A couple of interesting things can be seen here:
+
+1. There is an `Engine` API; more about this in the next section,
+
+2. Host functions are fully implemented in Go; `cgo` is no longer
+   necessary when declaring host functions. Moreover, host functions
+   no longer receive the famous `context` argument; more on that in a
+   second,
+
+3. Host functions can return multiple-values,
+
+4. Host functions can return any error _à la_ Go,
+
+5. The `ImportObject` API allows to register a collection of
+   “externals”: `Function`, `Memory` etc., i.e. any type that
+   implements the `IntoExtern` interface, to a given namespace (here
+   `math`),
+   
+6. The `Exports` API allows to read any kinds of externals, including
+   multiple memories.
+
+The entire API is more Go idiomatic, and very much simpler to use from
+our perspective.
+
+### A word about host functions
+
+Previously, a host function had to be declared with a `cgo` mapping:
+
+```go
+// extern int32_t sum(void *context, int32_t x, int32_t y);
+import "C"
+
+//export sum
+func sum(context unsafe.Pointer, x int32, y int32) int32 {
+	return x + y
+}
+```
+
+Now, a host function no longer needs `cgo`. That's a big
+user-experience improvement! A host function can be any Go functions
+or Go lambda functions.
+
+Also, a host function can now return an error, which was impossible
+before. It will automatically translate to a WebAssembly trap (with
+the new `Trap` API).
+
+Finally, a host function receive a slice of values and returns another
+slices of values, i.e. it supports multi-values!
```

Comment:
```suggestion
Finally, host functions receive and return slices of values, i.e. it supports
multi-values!
```

If this isn't always required you should specify that they "can" do this, as-is it sounds like all host functions must work this way

Reply:
It's always the case, it's the canonical form for host functions in Go now.
Thanks!